### PR TITLE
OHRM5X-2085: Handle Claim workflow states

### PR DESCRIPTION
**Pull request by DulsaraNethmin**

### CODE REVIEW

File: src/plugins/orangehrmClaimPlugin/Api/ClaimAttachmentAPI.php [hunk 1]
```diff
@@ -56,11 +59,21 @@ class ClaimAttachmentAPI extends Endpoint implements CrudEndpoint
...
     public const PARAMETER_ATTACHMENT_DESCRIPTION_MAX_LENGTH = 200;
+    public const PARAMETER_ALLOWED_ACTIONS = 'allowedActions';
+
+    public const ACTIONABLE_STATES_MAP = [
+        WorkflowStateMachine::CLAIM_ACTION_SUBMIT => 'SUBMIT',
+        WorkflowStateMachine::CLAIM_ACTION_APPROVE => 'APPROVE',
+        WorkflowStateMachine::CLAIM_ACTION_PAY => 'PAY',
+        WorkflowStateMachine::CLAIM_ACTION_CANCEL => 'CANCEL',
+        WorkflowStateMachine::CLAIM_ACTION_REJECT => 'REJECT'
+    ];
```
Comment (ManulMax) [inline src/plugins/orangehrmClaimPlugin/Api/ClaimAttachmentAPI.php:76]: seems like not using

**Review by ManulMax — CHANGES_REQUESTED** [review]

File: installer/Migration/V5_4_0/Migration.php [hunk 2]
```diff
@@ -301,6 +302,119 @@ public function up(): void
...
+            'SUBMITTED',
+            'SUPERVISOR',
+            WorkflowStateMachine::CLAIM_ACTION_REJECT,
+            'REJECTED',
+            '',
+            0
+        );
     }
 
     private function modifyClaimTables(): void
```
Comment (RajithaKumara) [inline installer/Migration/V5_4_0/Migration.php:420]: TODO::
change `ohrm_claim_request` status from TEXT to VARCHAR

File: installer/Migration/V5_4_0/Migration.php [hunk 3]
```diff
@@ -514,12 +633,12 @@ private function modifyDefaultRequiredPasswordStrength(): void
     }
 
     private function insertMenuItems(//TODO
-        string  $menu_title,
-        ?int    $screen_id,
-        ?int    $parent_id,
-        int     $level,
-        int     $order_hint,
-        int     $status,
+        string $menu_title,
```
Comment (RajithaKumara) [inline installer/Migration/V5_4_0/Migration.php:636]:
```suggestion
        string $menuTitle,
```
use camelCase

File: installer/Migration/V5_4_0/Migration.php [hunk 4]
```diff
@@ -684,4 +803,45 @@ private function createOAuth2Tables(): void
...
         $this->getConfigHelper()->setConfigValue('oauth.access_token_ttl', 'PT30M'); // 30 minutes
     }
+
+    private function insertWorkflowState(
+        int $workflow,
+        string $state,
+        string $role,
+        int $action,
+        string $resultingState,
+        string $rolesToNotify,
```
Comment (RajithaKumara) [inline installer/Migration/V5_4_0/Migration.php:813]: remove this parameter as all method call used empty string

File: installer/Migration/V5_4_0/Migration.php [hunk 5]
```diff
@@ -684,4 +803,45 @@ private function createOAuth2Tables(): void
...
+                    'roles_to_notify' => ':rolesToNotify',
+                    'priority' => ':priority',
+                ]
+            )
+            ->setParameter('workflow', $workflow)
+            ->setParameter('state', $state)
+            ->setParameter('role', $role)
+            ->setParameter('action', $action)
+            ->setParameter('resultingState', $resultingState)
+            ->setParameter('rolesToNotify', $rolesToNotify)
```
Comment (RajithaKumara) [inline installer/Migration/V5_4_0/Migration.php:834]:
```suggestion
            ->setParameter('rolesToNotify', '')
```

File: src/plugins/orangehrmClaimPlugin/entity/ClaimRequest.php [hunk 6]
```diff
@@ -104,11 +104,11 @@ class ClaimRequest
     private bool $isDeleted = false;
 
     /**
-     * @var string|null
+     * @var string
      *
      * @ORM\Column(name="status", type="string", length=255)
```
Comment (RajithaKumara) [inline src/plugins/orangehrmClaimPlugin/entity/ClaimRequest.php:109]: Modify migration to consistent with VARCHAR

File: src/plugins/orangehrmClaimPlugin/test/fixtures/ClaimAttachment.yaml [hunk 7]
```diff
@@ -32,14 +44,20 @@ CurrencyType:
...
+  3: { id: 4, emp_number: 4, added_by: 1, reference_id: 202301170000004, event_type_id: 2, description: "Salary expenses for temporary employee", currency_id: USD, is_deleted: false, status: 'SUBMITTED', created_date: "2022-09-01 08:00:00", submitted_date: "2022-09-30 18:00:00" },
+  4: { id: 5, emp_number: 4, added_by: 1, reference_id: 202301170000005, event_type_id: 3, description: "Expenses for social media advertising campaign", currency_id: CAD, is_deleted: false, status: 'REJECTED', created_date: "2022-08-15 12:00:00", submitted_date: "2022-08-20 16:00:00" },
+  5: { id: 6, emp_number: 1, added_by: 1, reference_id: 202301170000006, event_type_id: 4, description: "Training expenses for team building seminar", currency_id: USD, is_deleted: false, status: 'INITIATED', created_date: "2022-07-01 09:00:00", submitted_date: "2022-07-15 17:00:00" }
+  6: { id: 7, emp_number: 1, added_by: 1, reference_id: 202301170000007, event_type_id: 4, description: "Training expenses for team building seminar", currency_id: USD, is_deleted: false, status: 'REJECTED', created_date: "2022-07-01 09:00:00", submitted_date: "2022-07-15 17:00:00" }
 
 ClaimAttachment:
```
Comment (RajithaKumara) [inline src/plugins/orangehrmClaimPlugin/test/fixtures/ClaimAttachment.yaml:54]: in attachment table `eattach_attachment` column we are storing data as blobs, not as base64 encoded

**Review by RajithaKumara — APPROVED** [review]

**Pull request merged**